module_utils/foreman_helper.py

```diff
@@ -51,22 +53,189 @@ def filter_module_params(self):
     def get_server_params(self):
         return (self._foremanapi_server_url, self._foremanapi_username, self._foremanapi_password, self._foremanapi_verify_ssl)
 
+
+class ForemanAnsibleModule(ForemanBaseAnsibleModule):
+
+    def check_requirements(self):
+        if not HAS_NAILGUN:
+            self.fail_json(msg='The nailgun Python module is required',
+                           exception=NAILGUN_IMP_ERR)
+
     def connect(self, ping=True):
         entity_mixins.DEFAULT_SERVER_CONFIG = ServerConfig(
             url=self._foremanapi_server_url,
             auth=(self._foremanapi_username, self._foremanapi_password),
             verify=self._foremanapi_verify_ssl,
         )
 
+    def ping(self):
+        try:
+            return Ping().search_json()
+        except Exception as e:
+            self.fail_json(msg="Failed to connect to Foreman server: %s " % e)
+
+
+class ForemanApypieAnsibleModule(ForemanBaseAnsibleModule):
+
+    def check_requirements(self):
+        if not HAS_APYPIE:
+            self.fail_json(msg='The apypie Python module is required',
+                           exception=APYPIE_IMP_ERR)
+
+    def connect(self, ping=True):
+        try:
+            self.foremanapi = apypie.Api(uri=self._foremanapi_server_url,
+                                         username=self._foremanapi_username,
+                                         password=self._foremanapi_password,
+                                         api_version=2,
+                                         verify_ssl=self._foremanapi_verify_ssl)
+        except Exception as e:
+            self.fail_json(msg="Failed to connect to Foreman server: %s " % e)
+
         if ping:
             self.ping()
 
     def ping(self):
         try:
-            return Ping().search_json()
+            return self.foremanapi.resource('home').call('status')
         except Exception as e:
             self.fail_json(msg="Failed to connect to Foreman server: %s " % e)
 
+    def show_resource(self, resource, resource_id):
+        return self.foremanapi.resource(resource).call('show', {'id': resource_id})
+
+    def find_resource(self, resource, search, failsafe=False, thin=False):
+        response = self.foremanapi.resource(resource).call('index', {'search': search, 'thin': thin})
```

Comment:
One concern is we're now even more heavily dependent on scoped_search within Foreman, I think from https://community.theforeman.org/t/the-promise-of-a-stable-api/13119 it's clear that at least some of the dev team do not consider this an API. But I'm not sure what else we can do here?

Other than test against each server version and do releases with Foreman..

Reply:
Weren't we already with nailgun, as it does searches too?

Reply:
It was mixed, i personally tried to avoid it as most of the time there is a param you can use instead. Looking at https://github.com/theforeman/foreman-ansible-modules/blob/c6ba822834153576fcc2c598cd10bbd5ce3226ae/module_utils/ansible_nailgun_cement.py sometimes we use `search(query={'search': ` and sometimes we don't.

Reply:
But even if we don't pass a `query`, we always call `search`:
https://github.com/theforeman/foreman-ansible-modules/blob/c6ba822834153576fcc2c598cd10bbd5ce3226ae/module_utils/ansible_nailgun_cement.py#L266-L268

And this will do a scoped search in `nailgun`. I think the only possible way to avoid a scoped search is to pass an id.

Reply:
Well if you don't pass a query you are not depended on scoped search parameters as you're not using them.

Regardless, I don't see this as a big issue, marking as resolved. I'm happy to see how it goes over time.